htslib/sam.h

```diff
@@ -272,28 +289,336 @@ typedef struct {
+/*!
+ * @param type  Type of the searched line. Eg. "RG"
+ * @param id    Tag key defining the line. Eg. "ID"
+ * @param h     Hash set initialised by the caller with the values to be kept.
+ * @return      0 on success, -1 on failure
+ *
+ * Remove all lines of type <type> from the header, except the one
+ * specified in the hash set rh.
+ */
+int bam_hdr_remove_lines(bam_hdr_t *bh, const char *type, const char *id, void *h);
+
+/// Count the number of lines for a given header type
+/*!
+ * @param bh    BAM header
+ * @param type  Header type to count. Eg. "RG"
+ * @return  Number of lines of this type on success; -1 on failure
+ */
+int bam_hdr_count_lines(bam_hdr_t *bh, const char *type);
+
+/* ==== Key:val level methods ==== */
+
+/// Return the value associated with a key for a header line.
+/*!
+ * @param type      Type of the line to which the tag belongs. Eg. "SQ"
+ * @param ID_key    Tag key defining the line. Eg. "SN". Can be NULL, if looking for the first line.
+ * @param ID_value  Tag value associated with the key above. Eg. "ref1". Can be NULL, if ID_key is NULL.
+ * @param key       Key of the searched tag. Eg. "LN"
+ * @param ks        kstring where the value will be written
+ * @return          0 on success
+ *                 -1 if the requested tag does not exist
+ *                 -2 on other errors
+ *
+ * Looks for a specific key in a single SAM header line and writes the
+ * associated value into @p ks.  The header line is selected using the ID_key
+ * and ID_value parameters.  Any pre-existing content in @p ks will be
+ * overwritten.
+ */
+int bam_hdr_find_tag(bam_hdr_t *bh, const char *type, const char *ID_key, const char *ID_value, const char *key, kstring_t *ks);
+
+/// Remove the key from the line identified by type, ID_key and ID_value.
+/*!
+ * @param type      Type of the line to which the tag belongs. Eg. "SQ"
+ * @param ID_key    Tag key defining the line. Eg. "SN"
+ * @param ID_value  Tag value associated with the key above. Eg. "ref1"
+ * @param key       Key of the targeted tag. Eg. "M5"
+ * @return          1 if the key was removed; 0 if it was not present; -1 on error
+ */
+int bam_hdr_remove_tag(bam_hdr_t *bh, const char *type, const char *ID_key, const char *ID_value, const char *key);
+
+/// Get the target id for a given reference sequence name
+/*!
+ * @param ref  Reference name
+ * @return     Positive value on success, -1 if unknown reference
+ *
+ * Looks up a reference sequence by name in the reference hash table
+ * and returns the numerical target id.
+ */
+int bam_hdr_name2ref(bam_hdr_t *bh, const char *ref);
 
-    bam_hdr_t *bam_hdr_init(void);
-    bam_hdr_t *bam_hdr_read(BGZF *fp);
-    int bam_hdr_write(BGZF *fp, const bam_hdr_t *h) HTS_RESULT_USED;
-    void bam_hdr_destroy(bam_hdr_t *h);
-    int bam_name2id(bam_hdr_t *h, const char *ref);
-    bam_hdr_t* bam_hdr_dup(const bam_hdr_t *h0);
+/// Alias of bam_hdr_name2ref(), for backwards compatibility.
```

Comment:
We should probably ditch (deprecate) the use of `bam_hdr_t.sdict` altogether and make `bam_name2id` an actual alias of `bam_hdr_name2ref`. Currently, they have different implementations.